### PR TITLE
Closes #3916 WordFence background processes compatibility

### LINKED ISSUE #3916: Add auto-compatibility with WordFence for our background processes
**Describe the bug**
We have seen cases where WordFence is preventing our background processes from running correctly.

WordFence has a learning mode that can be used to allow the background processes. But It seems to be possible to programatically enable them by using some compatibility code. We need to explore this option to simplify our users experience.

**Backlog Grooming (for WP Media dev team use only)**
- [x] Reproduce the problem
- [x] Identify the root cause
- [x] Scope a solution
- [x] Estimate the effort

### PR DESCRIPTION
**Pull request by mostafa-hisham**
## Description

Create a new compatibility class `Engine\ThirdParty\Plugins\Security\WordFence`
to add or IP to WordFence whitelist
Fixes #3916 

## Type of change

- [x] Enhancement (non-breaking change which improves an existing functionality)

## Is the solution different from the one proposed during the grooming?
No

## How Has This Been Tested?

- [x] locally by checking if IP added in the WordFence IP whitelist
- [x] unit and integration tests

# Checklist:

Please delete the options that are not relevant.

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: inc/ThirdParty/Plugins/Security/WordFenceCompatibility.php [hunk 1]
```diff
@@ -0,0 +1,58 @@
+<?php
+namespace WP_Rocket\ThirdParty\Plugins\Security;
+
+use WP_Rocket\Event_Management\Subscriber_Interface;
+use Exception;
```
Comment (remyperona) [inline inc/ThirdParty/Plugins/Security/WordFenceCompatibility.php]:
```suggestion
```

File: inc/ThirdParty/Plugins/Security/WordFenceCompatibility.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+use wordfence;
+use WP_Rocket\Logger\Logger;
```
Comment (remyperona) [inline inc/ThirdParty/Plugins/Security/WordFenceCompatibility.php]:
```suggestion
```

File: inc/ThirdParty/Plugins/Security/WordFenceCompatibility.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	/**
+	 * Whitelist wp-rocket ips in wordfence firewall
+	 *
+	 * @since 3.10
+	 *
+	 * @return void
+	 */
+	public function whitelist_wordfence_firewall_ips() {
+
+		$ips = apply_filters( 'rocket_wordfence_whitelisted_ips', self::WHITELISTED_IPS );
```
Comment (remyperona) [inline inc/ThirdParty/Plugins/Security/WordFenceCompatibility.php:51]: Can you add a docblock for the filter?

File: tests/Fixtures/inc/ThirdParty/Plugins/Security/WordFence/wordfence.php [hunk 4]
```diff
@@ -0,0 +1,17 @@
+<?php
+/**
+ * Mocked WpeCommon for WP Engine to the minimum requirement for tests to run.
```
Comment (remyperona) [inline tests/Fixtures/inc/ThirdParty/Plugins/Security/WordFence/wordfence.php]:
```suggestion
 * Mocked wordfence to the minimum requirement for tests to run.
```

File: tests/Integration/inc/ThirdParty/Plugins/Security/WordFence/whitelistIPs.php [hunk 5]
```diff
@@ -0,0 +1,44 @@
+<?php
+
+namespace WP_Rocket\ThirdParty\Plugins\Security\WordFence;
+
+use WP_Rocket\Tests\Unit\TestCase;
```
Comment (remyperona) [inline tests/Integration/inc/ThirdParty/Plugins/Security/WordFence/whitelistIPs.php]:
```suggestion
use WP_Rocket\Tests\Integration\TestCase;
```

File: tests/Integration/inc/ThirdParty/Plugins/Security/WordFence/whitelistIPs.php [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+use WP_Rocket\ThirdParty\Plugins\Security\WordFenceCompatibility;
+use wordfence;
+use Mockery;
+use Brain\Monkey\Filters;
+use Brain\Monkey\Functions;
```
Comment (remyperona) [inline tests/Integration/inc/ThirdParty/Plugins/Security/WordFence/whitelistIPs.php]:
```suggestion
```

**Review by remyperona — CHANGES_REQUESTED** [review]

**Review by remyperona — APPROVED** [review]

**Comment by piotrbak** [comment]
@Tabrisrp @mostafa-hisham I'm not able to reproduce it. Even blocking the IP `135.125.83.227` in 3.9.4 version doesn't prevent RUCSS from working. Do we have any confirmation from support that whitelisting this IP in WordFence settings resolves the issue?

If so, then the IP is added to the whitelist correctly and that's what we could be testing further.

**Comment by piotrbak** [comment]
@Tabrisrp Is this one ready for QA? Should we try this approach? I don't have any feedback from support 😢

**Comment by remyperona** [comment]
@piotrbak Code-wise it's ok, but if we are not able to replicate the issue during QA, and we're not sure that's the right solution for the problem, we should wait.

**Review by piotrbak — APPROVED** [review]
@Tabrisrp as per talks with @GeekPress 

We're going to release it and see if new tickets are still coming, as we're not able to reproduce it.
[testrail-report-285.pdf](https://github.com/wp-media/wp-rocket/files/7410507/testrail-report-285.pdf)

**Pull request merged**